### ISSUE TITLE
Tips to speed up `<MSBuild/>` task calls?

### ISSUE THREAD

**Issue by jonathanpeppers** [dotnet/msbuild#8226]
### Issue Description

Android .NET 6+ apps can be built for 4 architectures, so we have a step that calls `<MSBuild/>` per-`$(RuntimeIdentifier)`:

https://github.com/xamarin/xamarin-android/blob/2a16828a6c73d837e81c3a2d0c1ae2840d00873e/src/Xamarin.Android.Build.Tasks/Microsoft.Android.Sdk/targets/Microsoft.Android.Sdk.AssemblyResolution.targets#L95-L100

Depending on the project, we see 150-300ms in just an extra `<MSBuild/>` task call. It doesn't appear that the `.targets` running inside the call are doing that much work.

Any thoughts on how we can speed this up?

![image](https://user-images.githubusercontent.com/840039/206013386-173dbc42-d463-4c29-b149-8a2bd86140de.png)

### Steps to Reproduce

1. Create a .NET MAUI project template
2. F5 on Android
3. Change some .xaml
4. F5 again, and look at the `.binlog`

### Data

[MauiApp43_Debug_AnyCPU_net7.0-android_Build_2022-12-06T14_11_51.1784775-06_00.binlog.zip](https://github.com/dotnet/msbuild/files/10170037/MauiApp43_Debug_AnyCPU_net7.0-android_Build_2022-12-06T14_11_51.1784775-06_00.binlog.zip)

### Versions & Configurations

In the above log I was using VS 2022 17.4.2 and .NET 7. I believe we've always noticed this since .NET 6, though.

### Regression?

No

/cc @rokonec

**Comment by jonathanpeppers** [comment]
The main important one is `RuntimeIdentifier=%(_RIDs.Identity)`, we are building an Android app for 4 RIDs and merging into a final `.apk` or `.aab` file.

**Comment by JanKrivanek** [comment]
@jonathanpeppers - We had a quick look with @rokonec and unfortunately there doesn't seem to be any hidden surprise - .csproj  needs re-evaluation that takes 66ms, build of the .csproj takes 86ms and rest to the 273ms of the MSBuild taks is taken up by numerous tasks pulled as dependencies or/and pre/after build tasks. So there unfortunately doesn't seem to be any low hanging fruit to tackle.

Though we'd like to provide deeper insight into your case - can you please provide us with ETW trace of the repro?
Please use the following command to ensure build events for proper correlation with binlog:

```
perfview collect /NoGui /Providers=*Microsoft-Build /BufferSize:8096 /CircularMB:8096 /NoNGenRundown /Merge:False /DataFile:rebuild-console-redacted-false
```

plus the binlog for the identical run.


Again - we are not promissing miracles - but at least we should be able to provide much deeper insight.

Thank you

**Comment by jonathanpeppers** [comment]
I had to put it in two files to get around GitHub's file size limit:

* [android-etw.zip](https://github.com/dotnet/msbuild/files/10269453/android-etw.zip)
* [android-etw2.zip](https://github.com/dotnet/msbuild/files/10269461/android-etw2.zip)

There are two `.binlogs` inside, the IDE does Build/Deploy in two steps. You would be interested in the first one (runs `Build`) where we call `<MSBuild/>`.

**Comment by JanKrivanek** [comment]
**tl;dr;:** Unfortunately no gaps identified in bottom-up-approach (answering "are there any ineficiencies or unneeded executions in the invocation path of the mentioned MSBuild task?"). Next step would be more holistic (and more complicated and involved) up-down-approach (answering "are involved maui build targets leveraging MSBuild properly or are there more efficient ways of achieving multitarget build with possible overlaps in targes evaluation?") - I'll try to summon someone more experienced to help me with this investigation (@rainersigwald - I hope to get your help here ;-))

---

Findings:

* Binlogging takes ~30% of execution time. Rest of investigation is after logging turned off (ETW only)
* Subtasks of MSBuild task takes about ~30% of execution time of the MSBuild task (no unexpected tasks nor duration here, I'm just clarifying the tasks without name)


Tasks | Sum   of DURATION_MSEC
-- | --
MSBuild | 303.156
ResolveAssemblyReference | 37.649
  | 35.408
ProcessFrameworkReferences | 15.831
ResolveRuntimePackAssets | 12.397
ResolvePackageAssets | 6.088
ResolveOverlappingItemGroupConflicts | 5.022
ResolvePackageFileConflicts | 4.021
ResolveTargetingPackAssets | 3.785
CheckForDuplicateItemMetadata | 3.699
GetPackageDirectory | 1.629


![image](https://user-images.githubusercontent.com/3809076/214124007-8384bcff-dd32-46f6-b94b-0d4ff7b71de2.png)


* The execution of the MSBuild task is ~45% evaluation, ~55% Build. Even when drilling very deep down the stack, there is no unexpected execution times or unexpected executions.

![image](https://user-images.githubusercontent.com/3809076/214123477-c8450581-23b0-4358-87f8-44bb6664d8cd.png)


* As already mentioned here - `AdditionalProperties="RuntimeIdentifier=%(_RIDs.Identity);$(_AdditionalProperties)"` - is causing the need to re-evaluate the project and rebuild the target - and currently results of previoius evaluation/build (with differen AdditionalPropertis) cannot be shared.
* **TBD:** Are there any viable improvements to `Microsoft.Android.Sdk.AssemblyResolution.targets` (to enable better up-to-date check; share portion of evaluation invoked by MSBuild task etc.)?

**Comment by jonathanpeppers** [comment]
The  `<MSBuild/>` call we are making results in populating a `@(ResolvedFileToPublish)` item group.

What is the best way to save an item group in a file (including all item metadata)? We could potentially use this file as an `Output` to skip the target in some cases. Incremental builds could load a file and populate an item group, skipping the `<MSBuild/>` call entirely.

**Comment by yuehuang010** [comment]
<Target /> have an Inputs and Outputs parameter and will skip the target if
all Outputs are newer than the Inputs.  You can use the <WriteLinesToFiles>
<https://learn.microsoft.com/en-us/visualstudio/msbuild/writelinestofile-task?view=vs-2022>
Task to dump the information.  Likewise, <ReadLinesFromFile> Task to absorb
it back.

Within the target, you can use <CallTarget> to invoke your Target that will
eventually call <MSBuild> task.

**Comment by jonathanpeppers** [comment]
@yuehuang010 how do you save item metadata with just `<WriteLinesToFiles/>` and `<ReadLinesFromFile/>`?

Is the only option a custom task?

**Comment by yuehuang010** [comment]
For metadata, a custom task is the only way to go.
Is it possible to avoid the metadata and only use the filepath?

**Comment by rainersigwald** [comment]
@jonathanpeppers how are you thinking of invalidating the incremental build? That is, under what circumstances would the MSBuild task need to be run when you had some cached results?

**Comment by jonathanpeppers** [comment]
The `Inputs` would probably be `$(MSBuildAllProjects)` and a file where we save important property values. I would need to think about & review other files.

It's possible this would create too many bugs. Luckily our CI has lots of MSBuild integration tests (even incremental builds) that would show if it is a bad idea or not.

**Comment by rokonec** [comment]
@JanKrivanek  FYI

**Comment by JanKrivanek** [comment]
Hi @jonathanpeppers - the next step is experimentation with `_ResolveAssemblies` target and direct dependencies (attempt to define inputs/outputs as files). Do you want to assign the item to yourself for the prototyping (we can then assist with review and measuring) or possibly close this investigation ticket? 
I'm as well open to any alternative proposals of next steps.

**Comment by jonathanpeppers** [comment]
We can track it here: https://github.com/xamarin/xamarin-android/issues/7775

It might be a little while until we get to this. Right now the huge push for us is fixing MAUI bugs.
